extmod/network_lwip.c

```diff
@@ -90,6 +100,241 @@ mp_obj_t mod_network_nic_ifconfig(struct netif *netif, size_t n_args, const mp_o
     }
 }
 
+mp_obj_t network_nic_config(struct netif *netif, size_t n_args, const mp_obj_t *args, mp_map_t *kwargs) {
+
+    if (kwargs->used == 0) {
+        // Get config value
+        if (n_args != 1) {
+            mp_raise_TypeError(MP_ERROR_TEXT("must query one param"));
+        }
+
+        switch (mp_obj_str_get_qstr(args[0])) {
+            case MP_QSTR_dhcp4: {
+                struct dhcp *dhcp = netif_dhcp_data(netif);
+                return dhcp != NULL && dhcp->state != DHCP_STATE_OFF ? mp_const_true : mp_const_false;
+            }
+            #if LWIP_IPV6_DHCP6
+            case MP_QSTR_dhcp6: {
+                struct dhcp6 *dhcp = netif_dhcp6_data(netif);
+                return dhcp != NULL && dhcp->state != DHCP6_STATE_OFF ? mp_const_true : mp_const_false;
+            }
+            #endif
+            #if LWIP_IPV6_AUTOCONFIG
+            case MP_QSTR_autoconf6: {
+                return netif->ip6_autoconfig_enabled ?  mp_const_true : mp_const_false;
+            }
+            #endif
+            case MP_QSTR_dns: {
+                const char *addr_str = ipaddr_ntoa(dns_getserver(0));
+                return mp_obj_new_str(addr_str, strlen(addr_str));
+            }
+            #if LWIP_VERSION_MAJOR >= 2
+            case MP_QSTR_prefer: {
+                return MP_OBJ_NEW_SMALL_INT(mp_mod_network_prefered_dns_ip);
+            }
+            #endif
+            case MP_QSTR_addr4: {
+                mp_obj_t tuple[2] = {
+                    netutils_format_ipv4_addr((uint8_t *)&netif->ip_addr, NETUTILS_BIG),
+                    netutils_format_ipv4_addr((uint8_t *)&netif->netmask, NETUTILS_BIG),
+                };
+                return mp_obj_new_tuple(2, tuple);
+            }
+            #if LWIP_IPV6
+            case MP_QSTR_addr6: {
+                mp_obj_t addrs[LWIP_IPV6_NUM_ADDRESSES];
+                size_t n_addrs = 0;
+                for (int i = 0; i < LWIP_IPV6_NUM_ADDRESSES; i++) {
+                    if (ip6_addr_isvalid(netif_ip6_addr_state(netif, i))) {
+                        const char *addr_str = ipaddr_ntoa(netif_ip_addr6(netif, i));
```

Comment:
Note that `ipaddr_ntoa` writes to a global static buffer... This is safe here because we immediately `mp_obj_new_str` it before calling again (but means this function is not thread safe).

If we use `ipaddr_ntoa_r` instead (with a stack-allocated buffer) then (assuming nothing else calls `ipaddr_ntoa`) we can save 56 bytes of bss by having the linker GC those functions.